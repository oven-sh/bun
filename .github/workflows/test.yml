name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
