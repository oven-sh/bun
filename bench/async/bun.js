import { bench, run } from "../runner.mjs";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
