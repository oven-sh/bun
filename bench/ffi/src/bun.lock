{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
