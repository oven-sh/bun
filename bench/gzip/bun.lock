{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
