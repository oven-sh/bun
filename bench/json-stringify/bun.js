import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
