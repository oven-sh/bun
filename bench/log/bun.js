import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
