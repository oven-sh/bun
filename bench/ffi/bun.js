import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
