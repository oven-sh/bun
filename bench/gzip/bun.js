import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
