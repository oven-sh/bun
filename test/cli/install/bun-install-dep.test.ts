import { npa } from "bun:internal-for-testing";
import { expect, test } from "bun:test";

const bitbucket = [
  "bitbucket:dylan-conway/public-install-test",
  "bitbucket.org:dylan-conway/public-install-test",
  "bitbucket.com:dylan-conway/public-install-test",
  "git@bitbucket.org:dylan-conway/public-install-test",
];

const tarball_remote = [
  "http://localhost:5000/no-deps/-/no-deps-2.0.0.tgz",
  "https://registry.npmjs.org/no-deps/-/no-deps-2.0.0.tgz",
];

const local_tarball = ["file:./path/to/tarball.tgz", "./path/to/tarball.tgz"];
const github = ["foo/bar"];
const folder = ["file:./path/to/folder"];

const gitlab = ["gitlab:dylan-conway/public-install-test", "gitlab.com:dylan-conway/public-install-test"];

const all = [
  "@scoped/package",
  "@scoped/package@1.0.0",
  "@scoped/package@1.0.0-beta.1",
  "@scoped/package@1.0.0-beta.1+build.123",
  "package",
  "package@1.0.0",
  "package@1.0.0-beta.1",
  "package@1.0.0-beta.1+build.123",
  ...bitbucket,
  ...github,
  ...gitlab,
  ...tarball_remote,
  ...local_tarball,
  ...github,
  "github:dylan-conway/public-install-test",
  "git@github.com:dylan-conway/public-install-test",
  "https://github.com/dylan-conway/public-install-test",
  "https://github.com/dylan-conway/public-install-test.git",
  "https://github.com/dylan-conway/public-install-test.git#semver:^1.0.0",
];

test.each(all)("npa %s", dep => {
  expect(npa(dep)).toMatchSnapshot();
});

const pkgJsonLike = [
  ["foo", "1.2.3"],
  ["foo", "latest"],
  ["foo", "workspace:*"],
  ["foo", "workspace:^1.0.0"],
  ["foo", "workspace:1.0.0"],
  ["foo", "workspace:1.0.0-beta.1"],
  ["foo", "workspace:1.0.0-beta.1+build.123"],
  ["foo", "workspace:1.0.0-beta.1+build.123"],
  ["foo", "workspace:1.0.0-beta.1+build.123"],
  ["bar", "^1.0.0"],
  ["bar", "~1.0.0"],
  ["bar", "> 1.0.0 < 2.0.0"],
  ["bar", "1.0.0 - 2.0.0"],
];

test.each(pkgJsonLike)('dependencies: {"%s": "%s"}', (name, version) => {
  expect(npa(name, version)).toMatchSnapshot();
});

test("bad", () => {
  expect(() => npa("-123!}{P}{!P#$s")).toThrow();
});
