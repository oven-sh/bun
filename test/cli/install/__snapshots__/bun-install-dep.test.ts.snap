// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`npa @scoped/package 1`] = `
{
  "name": "@scoped/package",
  "version": {
    "name": "@scoped/package",
    "tag": "latest",
    "type": "dist_tag",
  },
}
`;

exports[`npa @scoped/package@1.0.0 1`] = `
{
  "name": "@scoped/package",
  "version": {
    "alias": false,
    "name": "@scoped/package",
    "type": "npm",
    "version": "==1.0.0",
  },
}
`;

exports[`npa @scoped/package@1.0.0-beta.1 1`] = `
{
  "name": "@scoped/package",
  "version": {
    "alias": false,
    "name": "@scoped/package",
    "type": "npm",
    "version": "==1.0.0-beta.1",
  },
}
`;

exports[`npa @scoped/package@1.0.0-beta.1+build.123 1`] = `
{
  "name": "@scoped/package",
  "version": {
    "alias": false,
    "name": "@scoped/package",
    "type": "npm",
    "version": "==1.0.0-beta.1+build.123",
  },
}
`;

exports[`npa package 1`] = `
{
  "name": "package",
  "version": {
    "name": "package",
    "tag": "latest",
    "type": "dist_tag",
  },
}
`;

exports[`npa package@1.0.0 1`] = `
{
  "name": "package",
  "version": {
    "alias": false,
    "name": "package",
    "type": "npm",
    "version": "==1.0.0",
  },
}
`;

exports[`npa package@1.0.0-beta.1 1`] = `
{
  "name": "package",
  "version": {
    "alias": false,
    "name": "package",
    "type": "npm",
    "version": "==1.0.0-beta.1",
  },
}
`;

exports[`npa package@1.0.0-beta.1+build.123 1`] = `
{
  "name": "package",
  "version": {
    "alias": false,
    "name": "package",
    "type": "npm",
    "version": "==1.0.0-beta.1+build.123",
  },
}
`;

exports[`npa bitbucket:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "bitbucket:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa bitbucket.org:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "bitbucket.org:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa bitbucket.com:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "bitbucket.com:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa git@bitbucket.org:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "git@bitbucket.org:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa foo/bar 1`] = `
{
  "name": "",
  "version": {
    "owner": "foo",
    "ref": "",
    "repo": "bar",
    "type": "github",
  },
}
`;

exports[`npa gitlab:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "gitlab:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa gitlab.com:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "gitlab.com:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa http://localhost:5000/no-deps/-/no-deps-2.0.0.tgz 1`] = `
{
  "name": "",
  "version": {
    "name": "",
    "type": "tarball",
    "url": "http://localhost:5000/no-deps/-/no-deps-2.0.0.tgz",
  },
}
`;

exports[`npa https://registry.npmjs.org/no-deps/-/no-deps-2.0.0.tgz 1`] = `
{
  "name": "",
  "version": {
    "name": "",
    "type": "tarball",
    "url": "https://registry.npmjs.org/no-deps/-/no-deps-2.0.0.tgz",
  },
}
`;

exports[`npa file:./path/to/tarball.tgz 1`] = `
{
  "name": "",
  "version": {
    "name": "",
    "path": "./path/to/tarball.tgz",
    "type": "tarball",
  },
}
`;

exports[`npa ./path/to/tarball.tgz 1`] = `
{
  "name": "",
  "version": {
    "name": "",
    "path": "./path/to/tarball.tgz",
    "type": "tarball",
  },
}
`;

exports[`npa foo/bar 2`] = `
{
  "name": "",
  "version": {
    "owner": "foo",
    "ref": "",
    "repo": "bar",
    "type": "github",
  },
}
`;

exports[`npa github:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "dylan-conway",
    "ref": "",
    "repo": "public-install-test",
    "type": "github",
  },
}
`;

exports[`npa git@github.com:dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "",
    "repo": "git@github.com:dylan-conway/public-install-test",
    "type": "git",
  },
}
`;

exports[`npa https://github.com/dylan-conway/public-install-test 1`] = `
{
  "name": "",
  "version": {
    "owner": "dylan-conway",
    "ref": "",
    "repo": "public-install-test",
    "type": "github",
  },
}
`;

exports[`npa https://github.com/dylan-conway/public-install-test.git 1`] = `
{
  "name": "",
  "version": {
    "owner": "dylan-conway",
    "ref": "",
    "repo": "public-install-test",
    "type": "github",
  },
}
`;

exports[`npa https://github.com/dylan-conway/public-install-test.git#semver:^1.0.0 1`] = `
{
  "name": "",
  "version": {
    "owner": "",
    "ref": "semver:^1.0.0",
    "repo": "https://github.com/dylan-conway/public-install-test.git",
    "type": "git",
  },
}
`;

exports[`dependencies: {"foo": "1.2.3"} 1`] = `
{
  "alias": false,
  "name": "foo",
  "type": "npm",
  "version": "==1.2.3-foo",
}
`;

exports[`dependencies: {"foo": "latest"} 1`] = `
{
  "name": "foo",
  "tag": "latest",
  "type": "dist_tag",
}
`;

exports[`dependencies: {"foo": "workspace:*"} 1`] = `
{
  "name": "*foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"foo": "workspace:^1.0.0"} 1`] = `
{
  "name": "^1.0.0foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"foo": "workspace:1.0.0"} 1`] = `
{
  "name": "1.0.0foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"foo": "workspace:1.0.0-beta.1"} 1`] = `
{
  "name": "1.0.0-beta.1foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"foo": "workspace:1.0.0-beta.1+build.123"} 1`] = `
{
  "name": "1.0.0-beta.1+build.123foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"foo": "workspace:1.0.0-beta.1+build.123"} 2`] = `
{
  "name": "1.0.0-beta.1+build.123foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"foo": "workspace:1.0.0-beta.1+build.123"} 3`] = `
{
  "name": "1.0.0-beta.1+build.123foo",
  "type": "workspace",
}
`;

exports[`dependencies: {"bar": "^1.0.0"} 1`] = `
{
  "alias": false,
  "name": "bar",
  "type": "npm",
  "version": ">=1.0.0-bar <2.0.0",
}
`;

exports[`dependencies: {"bar": "~1.0.0"} 1`] = `
{
  "alias": false,
  "name": "bar",
  "type": "npm",
  "version": ">=1.0.0-bar <1.1.0",
}
`;

exports[`dependencies: {"bar": "> 1.0.0 < 2.0.0"} 1`] = `
{
  "alias": false,
  "name": "bar",
  "type": "npm",
  "version": ">1.0.0 && <2.0.0-bar",
}
`;

exports[`dependencies: {"bar": "1.0.0 - 2.0.0"} 1`] = `
{
  "alias": false,
  "name": "bar",
  "type": "npm",
  "version": ">=1.0.0 <=2.0.0-bar",
}
`;
